Add initial LoRa support

Mani-Sadhasivam opened the pull request:
Hello,

This is my second attempt on adding LoRa (Long Range low power) technology support to Zephyr. This initial PR includes only P2P (Peer to Peer) support without using the LoRaMac stack. As per the discussion on the previous PR #17685, I've made use of Semtech's [LoRaMac-node](https://github.com/Lora-net/LoRaMac-node/) repository for better code maintenance. I've imported both LoRaMac and LoRa drivers (SX1276) as a module [here](https://github.com/zephyrproject-rtos/loramac-node/pull/1) but only P2P related drivers are enabled for now.

Since this is an initial PR, I've left many TODO items in the Zephyr SX1276 Shim driver for tackling in future.

### What is working?

Only P2P support using SX1276 LoRa Radio driver. This has been tested using 96Boards Wistrio board. There is no frequency limitation as of now but only below bandwidths are supported:

```125KHz, 250KHz, 500KHz```

...and this limitation comes from Semtech's LoRa drivers.

### Future plans

LoRaWAN support

### Discussions

#5436 #5764 #17685

Thanks,
Mani

Mani-Sadhasivam committed:
include: Add initial LoRa API

Add initial LoRa API for P2P mode.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
drivers: lora: Add SX1276 LoRa Modem support

Add support for Semtech SX1276 LoRa Modem.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
boards: arm: 96b_wistrio: Add on-board SX1276 LoRa Modem suppport

Add support for Semtech SX1276 LoRa Modem found within the RAK811
module on the board.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
samples: drivers: Add LoRa sender sample

Add sample application for sending data packets over LoRa.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
samples: drivers: Add LoRa receiver sample

Add sample application for receiving data packets over LoRa.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
CODEOWNERS: Add entry for LoRa support

Add CODEOWNERS entry for LoRa API, drivers and samples.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
drivers: Add basic LoRa driver support

Add basic driver support for LoRa.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
west: Add LoRaMac-node module

Add LoRaMac-node module support to make use of Semtech LoRaWAN stack
and LoRa drivers.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam committed:
modules: Add LoRaMac Stack and drivers

Add LoRaMac module support for building the LoRaWAN stack and LoRa
drivers provided by Semtech.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam commented:
Tagging: @tbursztyka @alexanderwachter @jukkar @chris-makaio

KwonTae-young commented:
I have a custom board using the `STM32L151CBT` and `SX1276`.
I tested `sample/drivers/lora` and it works fine.
```
***** Booting Zephyr OS build v1.8.99-18626-gabe9fca35d9b *****
[00:00:00.222,000] <inf> sx1276: SX1276 Version:12 found
[00:00:05.749,000] <inf> lora_receive: Received data: helloworld
```

erwango commented on soc/arm/st_stm32/stm32l1/soc.h:
```diff
@@ -51,6 +51,12 @@
 #include <stm32l1xx_ll_i2c.h>
 #endif
 
+#if defined(CONFIG_RTC_STM32) || defined(CONFIG_COUNTER_RTC_STM32)
```
CONFIG_RTC_STM32 is no more supported, please remove

ulfalizer replied:
Could `drivers/counter/Kconfig.stm32_rtc` be removed too?

erwango replied:
Actually counter_rtc replaced rtc. rtc has been removed, but counter_rtc is still needed.

erwango commented on samples/drivers/lora/send/boards/96b_wistrio.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_SPI_STM32=y
```
This should not be needed.

erwango commented on samples/drivers/lora/receive/boards/96b_wistrio.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_SPI_STM32=y
```
ditto

erwango reviewed (dismissed):
Few comments after first quick review.
Can you split the PR and put SPI/RTC support on L1 in dedicated PRs so we can merge these first?

KwonTae-young commented on samples/drivers/lora/send/src/main.c:
```diff
@@ -0,0 +1,54 @@
+	if (ret < 0) {
+		LOG_ERR("LoRa send failed");
+		return;
+	}
```
I think send and recive should be repeated.

Mani-Sadhasivam replied:
Okay, we can use a loop count of 100 for this?

KwonTae-young commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,483 @@
+	Radio.Rx(0);
+
+	/* Wait until the data arrives */
+	k_sem_take(&dev_data.data_sem, K_FOREVER);
```
I think `sx1276_lora_recv()` needs timeout.

Mani-Sadhasivam replied:
@KwonTae-young Makes sense! I think the API should allow this timeout value to be specified by the application. If timeout passed is `0`, then it should block until the data arrives. What do you think?

Like below,

```
static int sx1276_lora_recv(struct device *dev, u8_t *data, u32_t timeout)
{
        int ret;

        Radio.SetMaxPayloadLength(MODEM_LORA, 255);
        Radio.Rx(0);

        if (!timeout) {
                /* Wait until the data arrives */
                k_sem_take(&dev_data.data_sem, K_FOREVER);
        } else {
                ret = k_sem_take(&dev_data.data_sem, K_MSEC(timeout));
                if (ret < 0) {
                        LOG_ERR("Receive timeout!");
                        return ret;
                }
        }

        memcpy(data, dev_data.rx_buf, dev_data.rx_len);

        return dev_data.rx_len;
}
```

alexanderwachter replied:
>I think the API should allow this timeout value to be specified by the application. If timeout passed is 0, then it should block until the data arrives. What do you think?

I would suggest using the already defined macros K_FOREVER and K_NO_WAIT for that purpose.

Mani-Sadhasivam replied:
> I would suggest using the already defined macros K_FOREVER and K_NO_WAIT for that purpose.

@alexanderwachter Okay. I think you're talking about passing `K_NO_WAIT` from the application instead of passing `0`.

KwonTae-young replied:
timeout works fine.
However, if the timeout occurs once in `samples/drivers/lora/receive`, it is returned and the program terminates.
```
***** Booting Zephyr OS version 2.0.99 *****              
[00:00:00.221,000] <inf> sx1276: SX1276 Version:12 found  
[00:00:01.228,000] <err> sx1276: Receive timeout!
[00:00:01.228,000] <err> lora_receive: LoRa receive failed
```

I am testing with the following modifications. (timeout = 1000ms)
``` diff
diff --git a/samples/drivers/lora/receive/src/main.c b/samples/drivers/lora/receive/src/main.c
index 46824dd69e..99b6985165 100644
--- a/samples/drivers/lora/receive/src/main.c
+++ b/samples/drivers/lora/receive/src/main.c
@@ -48,11 +48,10 @@ void main(void)
                len = lora_recv(lora_dev, data, 0);
                if (len < 0) {
                        LOG_ERR("LoRa receive failed");
-                       return;
+               } else {
+                       LOG_INF("Received data: %s", log_strdup(data));
                }
 
-               LOG_INF("Received data: %s", log_strdup(data));
-
                /* Receive data at 1s interval */
                k_sleep(1000);
        }

```
```
***** Booting Zephyr OS version 2.0.99 *****
[00:00:00.221,000] <inf> sx1276: SX1276 Version:12 found
[00:00:01.228,000] <err> sx1276: Receive timeout!
[00:00:01.228,000] <err> lora_receive: LoRa receive failed
[00:00:03.229,000] <err> sx1276: Receive timeout!
[00:00:03.229,000] <err> lora_receive: LoRa receive failed
[00:00:05.230,000] <err> sx1276: Receive timeout!
[00:00:05.230,000] <err> lora_receive: LoRa receive failed
...
...
```

Mani-Sadhasivam replied:
@KwonTae-young This is the expected behavior. And the termination depends on the user application. The sample application I have provided just terminates once the timeout is received.

KwonTae-young replied:
@Mani-Sadhasivam Okay. timeout works fine.

jukkar commented:
BTW, did anyone had plans to add LoRa support to the BSD socket layer after this initial support is merged?

Mani-Sadhasivam commented:
> BTW, did anyone had plans to add LoRa support to the BSD socket layer after this initial support is merged?

@jukkar Of course! I will work on it. That's would be the exact functionality required for FOTA.

Mani-Sadhasivam commented:
@erwango Created #19861 with SPI/RTC stuffs as you suggested. Once that gets merged, I'll rebase this PR.

Mani-Sadhasivam commented:
@erwango I'm not going to keep rebasing this PR until #19861 gets merged.

Mani-Sadhasivam commented:
@erwango PR rebased!

alexanderwachter commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,492 @@
+			LOG_ERR("Receive timeout!");
+			return ret;
+		}
+	}
```
I was thinking more about something like this:
```C
ret = k_sem_take(&dev_data.data_sem, timeout == K_FOREVER ? K_FOREVER :
                              timeout == K_NO_WAIT ? K_NO_WAIT : 
                              K_MSEC(timeout));
if (ret < 0) {
    LOG_ERR("Receive timeout!");
    return ret;
}
```
Otherwise, you would introduce a new value for K_FOREVER, which is -1.

Mani-Sadhasivam replied:
@alexanderwachter hmm, makes sense!

Mani-Sadhasivam replied:
@alexanderwachter Done!

alexanderwachter replied:
Thanks!

KwonTae-young commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,492 @@
+	Radio.Send(data, data_len);
+
+	return 0;
+}
```
Do you have a plan to support `LBT(Listen Before Talk)` features?

KwonTae-young commented:
@Mani-Sadhasivam Longer test times make the data strange.(`elloworldow`, `helloworldw`)
Is it okay for 96b_wistrio?

```
***** Booting Zephyr OS version 2.0.99 *****
[00:00:00.222,000] <inf> sx1276: SX1276 Version:12 found
[00:00:06.230,000] <inf> lora_receive: Received data: helloworld
[00:00:07.235,000] <inf> lora_receive: Received data: helloworld
[00:00:08.241,000] <inf> lora_receive: Received data: helloworld
[00:00:09.246,000] <inf> lora_receive: Received data: helloworld
[00:00:10.251,000] <inf> lora_receive: Received data: helloworld
[00:00:11.256,000] <inf> lora_receive: Received data: helloworld
[00:00:12.261,000] <inf> lora_receive: Received data: helloworld
[00:00:13.266,000] <inf> lora_receive: Received data: helloworld
[00:00:14.271,000] <inf> lora_receive: Received data: helloworld
[00:00:15.276,000] <inf> lora_receive: Received data: helloworld
[00:00:16.281,000] <inf> lora_receive: Received data: helloworld
[00:00:17.286,000] <inf> lora_receive: Received data: helloworld
[00:00:18.291,000] <inf> lora_receive: Received data: helloworld
[00:00:19.296,000] <inf> lora_receive: Received data: helloworld
[00:00:20.301,000] <inf> lora_receive: Received data: helloworld
[00:00:21.305,000] <inf> lora_receive: Received data: helloworld
[00:00:22.309,000] <inf> lora_receive: Received data: helloworld
[00:00:23.313,000] <inf> lora_receive: Received data: helloworld
[00:00:24.318,000] <inf> lora_receive: Received data: helloworld
[00:00:25.322,000] <inf> lora_receive: Received data: helloworld
[00:00:26.326,000] <inf> lora_receive: Received data: helloworld
[00:00:27.330,000] <inf> lora_receive: Received data: helloworld
[00:00:28.335,000] <inf> lora_receive: Received data: helloworld
[00:00:29.339,000] <inf> lora_receive: Received data: helloworld
[00:00:30.344,000] <inf> lora_receive: Received data: helloworld
[00:00:31.348,000] <inf> lora_receive: Received data: elloworldow
[00:00:32.352,000] <inf> lora_receive: Received data: elloworldow
[00:00:33.356,000] <inf> lora_receive: Received data: elloworldow
[00:00:34.361,000] <inf> lora_receive: Received data: elloworldow
[00:00:35.365,000] <inf> lora_receive: Received data: elloworldow
[00:00:36.369,000] <inf> lora_receive: Received data: elloworldow
[00:00:37.373,000] <inf> lora_receive: Received data: elloworldow
[00:00:39.381,000] <inf> lora_receive: Received data: elloworldow
[00:00:40.386,000] <inf> lora_receive: Received data: elloworldow
[00:00:41.391,000] <inf> lora_receive: Received data: elloworldow
[00:00:42.395,000] <inf> lora_receive: Received data: elloworldow
[00:00:44.404,000] <inf> lora_receive: Received data: elloworldow
[00:00:45.409,000] <inf> lora_receive: Received data: elloworldow
[00:00:46.413,000] <inf> lora_receive: Received data: elloworldow
[00:00:47.418,000] <inf> lora_receive: Received data: elloworldow
[00:00:48.422,000] <inf> lora_receive: Received data: elloworldow
[00:00:50.431,000] <inf> lora_receive: Received data: elloworldow
[00:00:51.436,000] <inf> lora_receive: Received data: elloworldow
[00:00:52.441,000] <inf> lora_receive: Received data: elloworldow
[00:00:53.445,000] <inf> lora_receive: Received data: elloworldow
[00:00:54.449,000] <inf> lora_receive: Received data: elloworldow
[00:00:55.454,000] <inf> lora_receive: Received data: elloworldow
[00:00:56.458,000] <inf> lora_receive: Received data: elloworldow
[00:00:57.463,000] <inf> lora_receive: Received data: elloworldow
[00:00:58.468,000] <inf> lora_receive: Received data: elloworldow
[00:00:59.473,000] <inf> lora_receive: Received data: elloworldow
[00:01:00.478,000] <inf> lora_receive: Received data: helloworldw
[00:01:01.482,000] <inf> lora_receive: Received data: helloworldw
[00:01:02.488,000] <inf> lora_receive: Received data: helloworldw
[00:01:03.492,000] <inf> lora_receive: Received data: helloworldw
[00:01:04.497,000] <inf> lora_receive: Received data: helloworldw
[00:01:05.502,000] <inf> lora_receive: Received data: helloworldw
[00:01:06.507,000] <inf> lora_receive: Received data: helloworldw
[00:01:07.511,000] <inf> lora_receive: Received data: helloworldw
[00:01:08.516,000] <inf> lora_receive: Received data: helloworldw
[00:01:09.521,000] <inf> lora_receive: Received data: helloworldw
[00:01:10.525,000] <inf> lora_receive: Received data: helloworldw
[00:01:11.530,000] <inf> lora_receive: Received data: helloworldw
[00:01:12.535,000] <inf> lora_receive: Received data: helloworldw
[00:01:13.539,000] <inf> lora_receive: Received data: helloworldw
[00:01:14.544,000] <inf> lora_receive: Received data: helloworldw
[00:01:15.549,000] <inf> lora_receive: Received data: helloworldw
[00:01:16.554,000] <inf> lora_receive: Received data: helloworldw
[00:01:17.559,000] <inf> lora_receive: Received data: helloworldw
[00:01:18.563,000] <inf> lora_receive: Received data: helloworldw
[00:01:19.567,000] <inf> lora_receive: Received data: helloworldw
[00:01:20.572,000] <inf> lora_receive: Received data: helloworldw
[00:01:21.576,000] <inf> lora_receive: Received data: helloworldw
[00:01:22.581,000] <inf> lora_receive: Received data: helloworldw
[00:01:23.585,000] <inf> lora_receive: Received data: helloworldw
[00:01:24.589,000] <inf> lora_receive: Received data: helloworldw
[00:01:25.593,000] <inf> lora_receive: Received data: helloworldw
[00:01:26.598,000] <inf> lora_receive: Received data: helloworldw
[00:01:27.602,000] <inf> lora_receive: Received data: helloworldw
[00:01:28.607,000] <inf> lora_receive: Received data: helloworldw
[00:01:29.611,000] <inf> lora_receive: Received data: helloworldw
[00:01:30.616,000] <inf> lora_receive: Received data: helloworldw
[00:01:31.621,000] <inf> lora_receive: Received data: helloworldw
[00:01:32.626,000] <inf> lora_receive: Received data: helloworldw
[00:01:33.631,000] <inf> lora_receive: Received data: helloworldw
[00:01:34.635,000] <inf> lora_receive: Received data: helloworldw
[00:01:35.640,000] <inf> lora_receive: Received data: helloworldw
[00:01:36.645,000] <inf> lora_receive: Received data: helloworldw
[00:01:37.649,000] <inf> lora_receive: Received data: helloworldw
[00:01:38.653,000] <inf> lora_receive: Received data: helloworldw
[00:01:39.658,000] <inf> lora_receive: Received data: helloworldw
[00:01:40.663,000] <inf> lora_receive: Received data: helloworldw
[00:01:41.668,000] <inf> lora_receive: Received data: helloworldw
[00:01:42.673,000] <inf> lora_receive: Received data: helloworldw
[00:01:43.677,000] <inf> lora_receive: Received data: helloworldw
[00:01:44.682,000] <inf> lora_receive: Received data: helloworldw
[00:01:45.686,000] <inf> lora_receive: Received data: helloworldw
[00:01:46.691,000] <inf> lora_receive: Received data: helloworldw
[00:01:47.695,000] <inf> lora_receive: Received data: helloworldw
[00:01:48.699,000] <inf> lora_receive: Received data: helloworldw
[00:01:49.704,000] <inf> lora_receive: Received data: helloworldw
[00:01:50.708,000] <inf> lora_receive: Received data: helloworldw
[00:01:51.713,000] <inf> lora_receive: Received data: elloworldow
[00:01:52.718,000] <inf> lora_receive: Received data: elloworldow
[00:01:53.723,000] <inf> lora_receive: Received data: elloworldow
[00:01:54.728,000] <inf> lora_receive: Received data: elloworldow
[00:01:55.732,000] <inf> lora_receive: Received data: elloworldow
[00:01:56.737,000] <inf> lora_receive: Received data: elloworldow
[00:01:57.743,000] <inf> lora_receive: Received data: elloworldow
[00:01:58.748,000] <inf> lora_receive: Received data: elloworldow
[00:01:59.753,000] <inf> lora_receive: Received data: elloworldow
[00:02:00.758,000] <inf> lora_receive: Received data: elloworldow
[00:02:01.763,000] <inf> lora_receive: Received data: elloworldow
[00:02:02.768,000] <inf> lora_receive: Received data: elloworldow
[00:02:03.774,000] <inf> lora_receive: Received data: elloworldow
[00:02:04.779,000] <inf> lora_receive: Received data: elloworldow
[00:02:05.784,000] <inf> lora_receive: Received data: elloworldow
[00:02:06.789,000] <inf> lora_receive: Received data: elloworldow
[00:02:07.794,000] <inf> lora_receive: Received data: elloworldow
[00:02:08.799,000] <inf> lora_receive: Received data: elloworldow
[00:02:09.805,000] <inf> lora_receive: Received data: elloworldow
[00:02:10.810,000] <inf> lora_receive: Received data: elloworldow
[00:02:11.816,000] <inf> lora_receive: Received data: elloworldow
[00:02:12.821,000] <inf> lora_receive: Received data: elloworldow
[00:02:13.827,000] <inf> lora_receive: Received data: elloworldow
[00:02:14.832,000] <inf> lora_receive: Received data: elloworldow
[00:02:15.838,000] <inf> lora_receive: Received data: elloworldow
[00:02:16.843,000] <inf> lora_receive: Received data: elloworldow
[00:02:17.849,000] <inf> lora_receive: Received data: helloworldw
[00:02:18.854,000] <inf> lora_receive: Received data: helloworldw
[00:02:19.860,000] <inf> lora_receive: Received data: helloworldw
[00:02:20.865,000] <inf> lora_receive: Received data: helloworldw
[00:02:21.871,000] <inf> lora_receive: Received data: helloworldw
```

Mani-Sadhasivam commented:
@KwonTae-young I can reproduce this on Wistrio board. Will dig into it. Thanks for noticing, I didn't run it too long to notice this error :(

alexanderwachter commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,492 @@
+	return 3;
+}
+
+void RtcSetAlarm(uint32_t timeout)
```
From rtc-board.h:
```C
 /*!
 * \brief Sets the alarm
 *
 * \note The alarm is set at now (read in this funtion) + timeout
 *
 * \param timeout [IN] Duration of the Timer ticks
 */
void RtcSetAlarm( uint32_t timeout );
```
I think you are mixint up ticks and us here.

Mani-Sadhasivam replied:
Ah, yes indeed! Will fix it.

alexanderwachter commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,492 @@
+	u8_t regval;
+
+	for (i = 0; i < size; i++) {
+		ret = sx1276_read(addr, &regval, 1);
```
Is there a reason why you are not using a burst read?

Mani-Sadhasivam replied:
This is not I2C! Is there something like this available for SPI too?

KwonTae-young replied:
Currently sx1276_write() and sx1276_read() operate by 1 byte.
I've only modified sx1276_read() like this:
``` diff
diff --git a/drivers/lora/sx1276.c b/drivers/lora/sx1276.c
index ae2e854..3b1cdfb 100644
--- a/drivers/lora/sx1276.c
+++ b/drivers/lora/sx1276.c
@@ -241,17 +241,12 @@ void SX1276WriteBuffer(u16_t addr, u8_t *buffer, u8_t size)
 
 void SX1276ReadBuffer(u16_t addr, u8_t *buffer, u8_t size)
 {
-       int ret, i;
-       u8_t regval;
-
-       for (i = 0; i < size; i++) {
-               ret = sx1276_read(addr, &regval, 1);
-               if (ret < 0) {
-                       LOG_ERR("Unable to read address: %x", addr);
-                       return;
-               }
+       int ret;
 
-               buffer[i] = regval;
+       ret = sx1276_read(addr, buffer, size);
+       if (ret < 0) {
+               LOG_ERR("Unable to read address: %x", addr);
+               return;
        }
 }
```

The figure below shows 10 bytes read of FIFO register.

**before**
![before](https://user-images.githubusercontent.com/10510127/67536617-f695df80-f712-11e9-894e-ff070d6d251c.png)

**after**
![after](https://user-images.githubusercontent.com/10510127/67536622-0281a180-f713-11e9-8ce5-745a63b77b14.png)

alexanderwachter replied:
@Mani-Sadhasivam you are still not using the burst mode. Is there a reason for that?

Mani-Sadhasivam replied:
@alexanderwachter Ah, I overlooked it. Will do.

Mani-Sadhasivam replied:
@alexanderwachter After doing burst read, the received data looks incorrect:

```
[00:00:01.129,000] <inf> lora_receive: Received data: hdmlnwosld
[00:00:03.135,000] <inf> lora_receive: Received data: hdmlnwosld
[00:00:05.141,000] <inf> lora_receive: Received data: hdmlnwosld
[00:00:07.147,000] <inf> lora_receive: Received data: hdmlnwosld
[00:00:09.153,000] <inf> lora_receive: Received data: hdmlnwosld
[00:00:11.159,000] <inf> lora_receive: Received data: hdmlnwosld
```
@KwonTae-young Did you also see the issue?

KwonTae-young replied:
@Mani-Sadhasivam Try adding `STM32_OSPEEDR_VERY_HIGH_SPEED` to the SPI pin configuration.

I tested it at https://github.com/KwonTae-young/zephyr/tree/lora_burst and it works fine.

``` diff
diff --git a/boards/arm/96b_wistrio/pinmux.c b/boards/arm/96b_wistrio/pinmux.c
index 12a11a30af14..19174cd0d4a4 100644
--- a/boards/arm/96b_wistrio/pinmux.c
+++ b/boards/arm/96b_wistrio/pinmux.c
@@ -27,9 +27,9 @@ static const struct pin_config pinconf[] = {
 	{STM32_PIN_PB9, STM32L1X_PINMUX_FUNC_PB9_I2C1_SDA},
 #endif /* CONFIG_I2C_1 */
 #ifdef CONFIG_SPI_1
-	{STM32_PIN_PA5, STM32L1X_PINMUX_FUNC_PA5_SPI1_SCK},
-	{STM32_PIN_PA6, STM32L1X_PINMUX_FUNC_PA6_SPI1_MISO},
-	{STM32_PIN_PA7, STM32L1X_PINMUX_FUNC_PA7_SPI1_MOSI},
+	{STM32_PIN_PA5, STM32L1X_PINMUX_FUNC_PA5_SPI1_SCK | STM32_OSPEEDR_VERY_HIGH_SPEED},
+	{STM32_PIN_PA6, STM32L1X_PINMUX_FUNC_PA6_SPI1_MISO | STM32_OSPEEDR_VERY_HIGH_SPEED},
+	{STM32_PIN_PA7, STM32L1X_PINMUX_FUNC_PA7_SPI1_MOSI | STM32_OSPEEDR_VERY_HIGH_SPEED},
 #endif /* CONFIG_SPI_1 */
 	/* RF_CTX_PA */
 	{STM32_PIN_PA4, STM32_PUSHPULL_PULLUP},
```

Mani-Sadhasivam replied:
@KwonTae-young Great! It works :) Btw, we only need the flag for SCLK.

KwonTae-young replied:
@Mani-Sadhasivam
I think we also need to add a burst write.
``` diff
diff --git a/drivers/lora/sx1276.c b/drivers/lora/sx1276.c
index d779025..5936e0f 100644
--- a/zephyrproject_lora3/zephyr/drivers/lora/sx1276.c
+++ b/zephyrproject_lora4/zephyr/drivers/lora/sx1276.c
@@ -221,21 +221,18 @@ int sx1276_read(u8_t reg_addr, u8_t *data, u8_t len)
 	return sx1276_transceive(reg_addr, false, data, len);
 }

-int sx1276_write(u8_t reg_addr, u8_t byte)
+int sx1276_write(u8_t reg_addr, u8_t *data, u8_t size)
 {
-	return sx1276_transceive(reg_addr | BIT(7), true, &byte, 1);
+	return sx1276_transceive(reg_addr | BIT(7), true, data, size);
 }

 void SX1276WriteBuffer(u16_t addr, u8_t *buffer, u8_t size)
 {
-	int ret, i;
+	int ret;

-	for (i = 0; i < size; i++) {
-		ret = sx1276_write(addr, buffer[i]);
-		if (ret < 0) {
-			LOG_ERR("Unable to write address: 0x%x", addr);
-			return;
-		}
+	ret = sx1276_write(addr, buffer, size);
+	if (ret < 0) {
+		LOG_ERR("Unable to write address: 0x%x", addr);
 	}
 }

@@ -308,13 +305,13 @@ void SX1276SetRfTxPower(int8_t power)
 	pa_config = (pa_config & RF_PACONFIG_OUTPUTPOWER_MASK) |
 			     ((power + 1) & 0x0F);
 #endif
-	ret = sx1276_write(SX1276_REG_PA_CONFIG, pa_config);
+	ret = sx1276_write(SX1276_REG_PA_CONFIG, &pa_config, 1);
 	if (ret < 0) {
 		LOG_ERR("Unable to write PA config");
 		return;
 	}

-	ret = sx1276_write(SX1276_REG_PA_DAC, pa_dac);
+	ret = sx1276_write(SX1276_REG_PA_DAC, &pa_dac, 1);
 	if (ret < 0) {
 		LOG_ERR("Unable to write PA dac");
 		return;
```

alexanderwachter commented:
Why are you using a dedicated RTC device instead of the zephyr timer API?

KwonTae-young commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,492 @@
+	dev_data.rx_len = size;
+
+	k_sem_give(&dev_data.data_sem);
+}
```
After Rx, it is still in Rx mode.
Even if i increase the delay from receive to 10s, the Rx interrupt continues to occur every 1 second on the receive board.
``` diff
diff --git a/samples/drivers/lora/receive/src/main.c b/samples/drivers/lora/receive/src/main.c
index c68e17ad34..8c396e67d8 100644
--- a/samples/drivers/lora/receive/src/main.c
+++ b/samples/drivers/lora/receive/src/main.c
@@ -54,6 +54,6 @@ void main(void)
                LOG_INF("Received data: %s", log_strdup(data));
 
                /* Receive data at 1s interval */
-               k_sleep(1000);
+               k_sleep(10000);
        }
 }
```

My board has the following symptoms.
**1. Empty data**
```
***** Booting Zephyr OS version 2.0.99 *****
[00:00:00.221,000] <inf> sx1276: SX1276 Version:12 found
[00:00:01.375,000] <inf> lora_receive: Received data: helloworld
[00:00:02.377,000] <inf> lora_receive: Received data: helloworld
[00:00:03.379,000] <inf> lora_receive: Received data:
***** Booting Zephyr OS version 2.0.99 *****
[00:00:00.221,000] <inf> sx1276: SX1276 Version:12 found
[00:00:01.200,000] <inf> lora_receive: Received data: helloworld
[00:00:02.202,000] <inf> lora_receive: Received data:
***** Booting Zephyr OS version 2.0.99 *****
```

**2. SPI RXNE loop**
In this case, the board stops working.
```
(gdb) bt
#0  0x0800798a in LL_SPI_IsActiveFlag_RXNE (SPIx=0x40013000)
    at /work/zephyrproject_lora2/modules/hal/stm32/stm32cube/stm32l1xx/drivers/include/stm32l1xx_ll_spi.h:739
#1  ll_func_rx_is_not_empty (spi=0x40013000)
    at /work/zephyrproject_lora2/zephyr/drivers/spi/spi_ll_stm32.h:40
#2  spi_stm32_shift_m (data=0x20001e50 <spi_stm32_dev_data_1>, spi=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/drivers/spi/spi_ll_stm32.c:124
#3  spi_stm32_shift_frames (data=0x20001e50 <spi_stm32_dev_data_1>, spi=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/drivers/spi/spi_ll_stm32.c:187
#4  transceive (asynchronous=false, signal=0x0, rx_bufs=<optimized out>, 
    tx_bufs=0x20000bb0 <z_main_stack+584>, config=0x20000200 <dev_data+12>, 
    dev=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/drivers/spi/spi_ll_stm32.c:439
#5  spi_stm32_transceive (dev=<optimized out>, config=0x20000200 <dev_data+12>, 
    tx_bufs=0x20000bb0 <z_main_stack+584>, rx_bufs=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/drivers/spi/spi_ll_stm32.c:462
#6  0x0800565a in z_impl_spi_transceive (config=0x20000200 <dev_data+12>, 
    rx_bufs=0x20000bb8 <z_main_stack+592>, tx_bufs=0x20000bb0 <z_main_stack+584>, 
    dev=<optimized out>) at /work/zephyrproject_lora2/zephyr/include/drivers/spi.h:263
#7  spi_transceive (config=0x20000200 <dev_data+12>, 
--Type <RET> for more, q to quit, c to continue without paging--
    rx_bufs=0x20000bb8 <z_main_stack+592>, tx_bufs=0x20000bb0 <z_main_stack+584>, 
    dev=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/samples/drivers/lora/receive/build/zephyr/include/generated/syscalls/spi.h:33
#8  sx1276_transceive (length=<optimized out>, data=<optimized out>, write=false, 
    reg=<optimized out>) at /work/zephyrproject_lora2/zephyr/drivers/lora/sx1276.c:212
#9  sx1276_read (reg_addr=reg_addr@entry=1 '\001', 
    data=data@entry=0x20000bdf <z_main_stack+631> " \005", len=len@entry=1 '\001')
    at /work/zephyrproject_lora2/zephyr/drivers/lora/sx1276.c:221
#10 0x08005684 in SX1276ReadBuffer (addr=<optimized out>, 
    buffer=0x20000bff <z_main_stack+663> "", size=<optimized out>)
    at /work/zephyrproject_lora2/zephyr/drivers/lora/sx1276.c:248
#11 0x0800951c in SX1276Read (addr=addr@entry=1)
    at /work/zephyrproject_lora2/modules/lib/loramac-node/radio/sx1276/sx1276.c:1267
#12 0x08009538 in SX1276SetOpMode (opMode=<optimized out>)
    at /work/zephyrproject_lora2/modules/lib/loramac-node/radio/sx1276/sx1276.c:1219
#13 0x08006094 in SX1276SetRx (timeout=timeout@entry=0)
    at /work/zephyrproject_lora2/modules/lib/loramac-node/radio/sx1276/sx1276.c:1035
#14 0x0800542a in sx1276_lora_recv (dev=<optimized out>, data=<optimized out>, timeout=-1)
    at /work/zephyrproject_lora2/zephyr/drivers/lora/sx1276.c:355
--Type <RET> for more, q to quit, c to continue without paging--
#15 0x080044e0 in lora_recv (timeout=-1, data=0x20000c40 <z_main_stack+728> "helloworld", 
    dev=0x20001fa0 <__device_sx1276_lora>)
    at /work/zephyrproject_lora2/zephyr/include/drivers/lora.h:133
#16 main () at /work/zephyrproject_lora2/zephyr/samples/drivers/lora/receive/src/main.c:48
(gdb) 
```

I consulted the [ping-pong](https://github.com/Lora-net/LoRaMac-node/blob/develop/src/apps/ping-pong/NucleoL152/main.c) source of `loramac-node` and found out that **`Sleep()`** is needed.
https://github.com/Lora-net/LoRaMac-node/blob/develop/src/apps/ping-pong/NucleoL152/main.c#L350-L364
``` c
void OnTxDone( void )
{
    Radio.Sleep( );
    State = TX;
}

void OnRxDone( uint8_t *payload, uint16_t size, int16_t rssi, int8_t snr )
{
    Radio.Sleep( );
    BufferSize = size;
    memcpy( Buffer, payload, BufferSize );
    RssiValue = rssi;
    SnrValue = snr;
    State = RX;
}
```


I think the source should be added as below.
``` diff
diff --git a/drivers/lora/sx1276.c b/drivers/lora/sx1276.c
index aedd0d841f..11bd93033c 100644
--- a/drivers/lora/sx1276.c
+++ b/drivers/lora/sx1276.c
@@ -327,6 +327,11 @@ void SX1276SetRfTxPower(int8_t power)
        }
 }
 
+static void sx1276_tx_done(void)
+{
+       Radio.Sleep();
+}
+
 static int sx1276_lora_send(struct device *dev, u8_t *data, u32_t data_len)
 {
        Radio.SetMaxPayloadLength(MODEM_LORA, data_len);
@@ -338,6 +343,8 @@ static int sx1276_lora_send(struct device *dev, u8_t *data, u32_t data_len)
 
 static void sx1276_rx_done(u8_t *payload, u16_t size, int16_t rssi, int8_t snr)
 {
+       Radio.Sleep();
+
        dev_data.rx_buf = payload;
        dev_data.rx_len = size;
 
@@ -472,6 +479,7 @@ static int sx1276_lora_init(struct device *dev)
 
        k_sem_init(&dev_data.data_sem, 0, UINT_MAX);
 
+       dev_data.sx1276_event.TxDone = sx1276_tx_done;
        dev_data.sx1276_event.RxDone = sx1276_rx_done;
        Radio.Init(&dev_data.sx1276_event);
```

console output
```
***** Booting Zephyr OS version 2.0.99 *****
[00:00:00.222,000] <inf> sx1276: SX1276 Version:12 found
[00:00:01.025,000] <inf> lora_receive: Received data: helloworld
[00:00:03.024,000] <inf> lora_receive: Received data: helloworld
[00:00:05.025,000] <inf> lora_receive: Received data: helloworld
...
...
[01:00:54.229,000] <inf> lora_receive: Received data: helloworld
[01:00:56.228,000] <inf> lora_receive: Received data: helloworld
```

[Symptoms with strange data](https://github.com/zephyrproject-rtos/zephyr/pull/18998#issuecomment-544810918) were also fixed by this fix.

Mani-Sadhasivam replied:
@KwonTae-young Great! Thanks for looking into this. I still haven't found time to look into this. Will test this change and integrate it very soon.

Thanks once again :)

Mani-Sadhasivam commented:
> Why are you using a dedicated RTC device instead of the zephyr timer API?

That's because, the API calls for it explicitly!

Mani-Sadhasivam commented:
@KwonTae-young Tested and incorporated your changes.

tbursztyka commented on include/drivers/lora.h:
```diff
@@ -0,0 +1,136 @@
+	BW_500_KHZ,
+};
+
+enum lora_spreading_factor {
```
why not using original's lora node API naming, so datarate: enum  lora_datarate ?
(Maybe that would be simpler for those who know/use lora node already)

Mani-Sadhasivam replied:
Hmm. I thought of following the exact convention used in hardware but you're right, it might be easier if we use 'datarate' for lora-node users.

tbursztyka commented on include/drivers/lora.h:
```diff
@@ -0,0 +1,136 @@
+	int preamble_len;
+	int coding_rate;
+	int tx_power;
+	bool tx;
```
to keep proper struct alignement after the type changes I noted (unless you have good reasons to keep what you have), you could move attributes so premble_len goes before tx_power. etc...

Mani-Sadhasivam replied:
ok

tbursztyka commented on drivers/CMakeLists.txt:
```diff
@@ -35,6 +35,7 @@ add_subdirectory_if_kconfig(hwinfo)
 add_subdirectory_if_kconfig(espi)
```
maybe rework the commit title: drivers: Add LoRa driver directory

Though you could merge this with the actual driver as well. As you like.

Mani-Sadhasivam replied:
> maybe rework the commit title: drivers: Add LoRa driver directory

okay.

> Though you could merge this with the actual driver as well. As you like.

The reasoning behind splitting the commits is, it will be easier for other developers to look into the actual commit which adds the driver. They may use that commit as a reference for adding support for some other lora modems.

tbursztyka replied:
ok

tbursztyka reviewed (commented):
Haven't much to say. It's nice to see a proper PR for adding LoRa into Zephyr.

Just a few comments on the API config.

alexanderwachter commented:
> > Why are you using a dedicated RTC device instead of the zephyr timer API?
> 
> That's because, the API calls for it explicitly!

The zephyr timer API is an RTC too, but it has a list of timeout values. I'm ok with using a dedicated RTC, but the timer would be much easier.

Mani-Sadhasivam commented on include/drivers/lora.h:
```diff
@@ -0,0 +1,136 @@
+struct lora_modem_config {
+	u32_t frequency;
+	int bandwidth;
+	int spreading_factor;
```
ok

Mani-Sadhasivam commented on include/drivers/lora.h:
```diff
@@ -0,0 +1,136 @@
+
+struct lora_modem_config {
+	u32_t frequency;
+	int bandwidth;
```
ok

Mani-Sadhasivam commented on include/drivers/lora.h:
```diff
@@ -0,0 +1,136 @@
+	int bandwidth;
+	int spreading_factor;
+	int preamble_len;
+	int coding_rate;
```
ok

Mani-Sadhasivam commented on include/drivers/lora.h:
```diff
@@ -0,0 +1,136 @@
+	u32_t frequency;
+	int bandwidth;
+	int spreading_factor;
+	int preamble_len;
```
Ack

Mani-Sadhasivam commented on include/drivers/lora.h:
```diff
@@ -0,0 +1,136 @@
+	int spreading_factor;
+	int preamble_len;
+	int coding_rate;
+	int tx_power;
```
Actually, `s8_t`.

tbursztyka replied:
true

Mani-Sadhasivam commented:
> The zephyr timer API is an RTC too, but it has a list of timeout values. I'm ok with using a dedicated RTC, but the timer would be much easier.

I can see only the system timer API. Which one are you referring to?

Mani-Sadhasivam commented:
@tbursztyka Thanks for the review! I agree with the API changes you requested. The fact that Zephyr's LoRa port should align with what the community is used to, makes much sense. Pushed the rebased changes.

alexanderwachter commented:
> > The zephyr timer API is an RTC too, but it has a list of timeout values. I'm ok with using a dedicated RTC, but the timer would be much easier.
> 
> I can see only the system timer API. Which one are you referring to?

You can either use the timeout API from timeout_q.h, which is the API the kernel uses (very low overhead) or the timer API exposed to the applications [1]

[1] https://docs.zephyrproject.org/latest/reference/kernel/timing/timers.html

Mani-Sadhasivam commented:
> > > The zephyr timer API is an RTC too, but it has a list of timeout values. I'm ok with using a dedicated RTC, but the timer would be much easier.
> > 
> > 
> > I can see only the system timer API. Which one are you referring to?
> 
> You can either use the timeout API from timeout_q.h, which is the API the kernel uses (very low overhead) or the timer API exposed to the applications [1]
> 
> [1] https://docs.zephyrproject.org/latest/reference/kernel/timing/timers.html

@alexanderwachter [1] is based on the system clock, so it is not RTC. I think it is fine to explicitly depend on the rtc/counter API since it is there for that purpose.

alexanderwachter commented:
@Mani-Sadhasivam sure you can stick to the counter API. It was just a suggestion to not depend on an RTC. As far as I can see, the only thing the RTC alarm does is provide timeouts with a precision of one ms. This could easily be achieved using the timeout API.

Mani-Sadhasivam commented:
> @Mani-Sadhasivam sure you can stick to the counter API. It was just a suggestion to not depend on an RTC. As far as I can see, the only thing the RTC alarm does is provide timeouts with a precision of one ms. This could easily be achieved using the timeout API.

@alexanderwachter Yes, right now the use case of RTC API is limited. But we should understand that this is just an intial PR adding the LoRa support. When we add support for LoRaWAN, we need features like calender etc... (required by LoRaMAC). So, considering that I'll just stick to this.

Thanks for your suggestion!

tbursztyka reviewed (approved):
No other comments,  it seems fine to me.

Mani-Sadhasivam commented:
@wentongwu @alexanderwachter @jukkar @KwonTae-young Can you please review this PR?

Mani-Sadhasivam committed:
boards: arm: 96b_wistrio: Use STM32_OSPEEDR_VERY_HIGH_SPEED for SPI1_SCK

Add Add STM32_OSPEEDR_VERY_HIGH_SPEED flag for SPI1_SCK to function
properly. This is needed for the proper communication with the LoRa
modem. Without this flag, the received data is mangled when burst
read is performed.

Signed-off-by: Manivannan Sadhasivam <mani@kernel.org>

Mani-Sadhasivam commented:
@alexanderwachter Implemented burst read. Please take a look.

alexanderwachter commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,494 @@
+
+extern DioIrqHandler *DioIrq[];
+
+int sx1276_dio_pins[SX1276_MAX_DIO] = {
```
Can you make this const?

jukkar replied:
Shouldn't sx1276_dio_pins be static?

alexanderwachter commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,494 @@
+	DT_INST_0_SEMTECH_SX1276_DIO_GPIOS_PIN_4,
+};
+
+static char sx1276_dio_ports[SX1276_MAX_DIO][6] = {
```
Can you make this const?

alexanderwachter commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,494 @@
+{
+	int ret, i;
+
+	for (i = 0; i < size; i++) {
```
Please use the bulk transfer here. Same as the `SX1276ReadBuffer`

Mani-Sadhasivam replied:
Done, thanks

jhedberg commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,491 @@
+	const struct spi_buf buf[2] = {
+		{
+			.buf = &reg,
+			.len = 1
```
`sizeof(reg)`

jhedberg commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,491 @@
+	};
+	struct spi_buf_set tx = {
+		.buffers = buf,
+		.count = 2,
```
`ARRAY_SIZE(buf)`

jhedberg commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,491 @@
+	if (!write) {
+		const struct spi_buf_set rx = {
+			.buffers = buf,
+			.count = 2
```
`ARRAY_SIZE(buf)`

jhedberg reviewed (commented):
A couple of small change suggestions to get rid of magic numbers

Mani-Sadhasivam commented:
pinging @galak

Mani-Sadhasivam commented:
@carlescufi The `board.h` definition in loramac-node lib conflicts with the zephyr `board.h` file. What should we do to fix it? Change the include path of `board.h` in loramac-node so that it can become `boards/board.h`?

Mani-Sadhasivam commented:
@galak May I know when this PR can be merged? So far most of the reviewers are happy with this and have their review in place. I do not want to keep rebasing this PR and FYI I already got LoRaWAN in the works. Only issue now is the CI failure for which I have asked a question above on how to fix it!

carlescufi commented:
> @carlescufi The `board.h` definition in loramac-node lib conflicts with the zephyr `board.h` file. What should we do to fix it? Change the include path of `board.h` in loramac-node so that it can become `boards/board.h`?

So if I understand correctly [this board.h](https://github.com/zephyrproject-rtos/loramac-node/blob/master/boards/board.h) conflicts with the `board.h` that some of Zephyr's boards contain? I think that those `board.h` in Zephyr are no longer really necessary if you use Device Tree, but nonetheless they are there. I suggest that you apply your proposal and include loramac-node from the root of the tree, meaning `boards/board.h`, since that has no chance of colliding with Zephyr.

carlescufi commented:
@Mani-Sadhasivam have you checked the shippable failure?

carlescufi commented:
@erwango could you please re-review?

erwango commented on samples/drivers/lora/send/boards/96b_wistrio.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_SPI_1=y
```
Can't we put this directly in board's .defconfig under `if LORA`?

Mani-Sadhasivam replied:
IIUC, it is the duty of sample applications to enable the drivers needed for the functionality and board defconfig should only enable the bare minimum ones needed for basic usage. Am I missing something?

erwango replied:
This is true, but doesn't prevent to do some conditional configuration in the board itself when we know it'll be always valid (part of the board BSP).
So in the board something like:
```
if LORA
config SPI_1
     default  y
```
would be always valid and do not enable stuff when not required.
It is still application responsibility to enable `LORA` and `SPI_1` is not enabled when LORA is not required (Non-LORA application for instance).

erwango commented on samples/drivers/lora/receive/boards/96b_wistrio.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_SPI_1=y
```
in boards .defconfig ?

erwango reviewed (dismissed):
Minor comment

alexanderwachter commented:
@Mani-Sadhasivam @carlescufi we should build loramac-node and the lora-driver as a library. I did some tests and it worked for me.
I used
```
zephyr_library_named(loramac-node)

zephyr_library_sources_ifdef(CONFIG_LORA_SX1276 sx1276.c)
```
in the zephyr driver and
```
set(ZEPHYR_CURRENT_LIBRARY loramac-node)

zephyr_library_include_directories(boards)
zephyr_library_include_directories(mac)
zephyr_library_include_directories(radio)
zephyr_library_include_directories(.)
```
additionally I used `zephyr_library_sources_ifdef` for the source files.

`set(ZEPHYR_CURRENT_LIBRARY loramac-node)` is maybe a bit hacky. @nashif What is the best way to build a library with a driver and a module?

carlescufi commented:
@SebastianBoe could you please take a look at [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/18998#issuecomment-566222140)?

SebastianBoe commented:
@alexanderwachter : Why do you want/need the two sets of sources to be in the same library?

SebastianBoe commented:
Doing

set(ZEPHYR_CURRENT_LIBRARY loramac-node)

is not too hacky. It is a public part of the API.

alexanderwachter commented:
> @alexanderwachter : Why do you want/need the two sets of sources to be in the same library?

@SebastianBoe The driver part is kind of a zephyr wrapper for the LoRaMAC-node library.
It needs to import the API from the lib. I thought it's a good idea to build them in one library and only expose the driver to Zephyr. It solves the redefinition issue of some utility functions like the MAX() macro.

SebastianBoe commented:
You say "we should build loramac-node and the lora-driver as a library." can you mention why and how this compares to what is done now?

Also, the loramac-node PR is very big, could you show where the code you want to change is?

alexanderwachter commented:
The relevant part is in https://github.com/zephyrproject-rtos/loramac-node/blob/master/CMakeLists.txt and https://github.com/zephyrproject-rtos/zephyr/pull/18998/files#diff-4d3f1718e6c473abc54edf7fd4c5830d

I'm not a CMake expert, but as far as I understand, we add the include directories to the whole Zephyr build instead of limiting it to the scope of building the LoRAMAC-node object.

SebastianBoe commented:
I see. Then

`set(ZEPHYR_CURRENT_LIBRARY loramac-node)
`
should be fine. I can't see any issues with it at least. If it does cause problems you could alternatively define an INTERFACE library that just has encoded the include directories in it, and link each library that needs them with this library.

alexanderwachter commented:
@Mani-Sadhasivam can you change the CMake files to build a library?
If you need help, you can ping me and I'll prepare a PR for the loramac-node repo.

carlescufi reviewed (changes requested):
I agree with @alexanderwachter 's remarks. Also, I believe there are too many changes to the upstream loramac-node repository. @Mani-Sadhasivam can we find a way to build this as a library as suggested by @alexanderwachter and also to try to reduce the changes introduced to the upstream project?

carlescufi commented:
> I agree with @alexanderwachter 's remarks. Also, I believe there are too many changes to the upstream loramac-node repository. @Mani-Sadhasivam can we find a way to build this as a library as suggested by @alexanderwachter and also to try to reduce the changes introduced to the upstream project?

To follow-up on this @Mani-Sadhasivam , since you have been blocked for a while now. Can you please post a diff of the contents of the current zephyrproject-rtos fork of the loramac-node repo with the upstream, and then add this as a separate commit? i.e. I think we should rewrite the history of [the fork](https://github.com/zephyrproject-rtos/loramac-node) so that it's a clean import + any addiitonal changes required by Zephyr itself. Sorry I just found out about this. I'll help out during the process.

Mani-Sadhasivam commented:
> > I agree with @alexanderwachter 's remarks. Also, I believe there are too many changes to the upstream loramac-node repository. @Mani-Sadhasivam can we find a way to build this as a library as suggested by @alexanderwachter and also to try to reduce the changes introduced to the upstream project?
> 
> To follow-up on this @Mani-Sadhasivam , since you have been blocked for a while now. Can you please post a diff of the contents of the current zephyrproject-rtos fork of the loramac-node repo with the upstream, and then add this as a separate commit? i.e. I think we should rewrite the history of [the fork](https://github.com/zephyrproject-rtos/loramac-node) so that it's a clean import + any addiitonal changes required by Zephyr itself. Sorry I just found out about this. I'll help out during the process.

@carlescufi As per the conversation with @alexanderwachter, he will try to come up with ways to minimalize the diff stat with upstream repo. As said above, my time is very limited these days and I'd prefer to have the community contributions for this effort :)

alexanderwachter commented on drivers/lora/sx1276.c:
```diff
@@ -0,0 +1,491 @@
+		return ret;
+	}
+
+	memcpy(data, dev_data.rx_buf, dev_data.rx_len);
```
Copying the memory here might cause problems. The pointer points to a global buffer defined in the LoRa stack radio.c (`uint8_t RadioRxPayload[255];`). If we are not fast enough to copy the data, it might get overwritten.
An option would be to either have a message queue with N buffers.

jukkar replied:
The driver could utilize net_buf here as is done by Bluetooth and networking subsystems, instead of using one local buffer.

alexanderwachter replied:
The buffer is defined in the external module. We should use it as it is.

Mani-Sadhasivam replied:
@alexanderwachter Is it a good idea to share the buffer to application? Wouldn't it be a problem if userspace split is implemented?

alexanderwachter replied:
We should not share the buffer with the user.
I meant we should use the implementation as provided.

carlescufi commented:
@rlubos would you mind reviewing this PR alongside this commit in the module:

https://github.com/zephyrproject-rtos/loramac-node/commit/ed2c1b3ba6f3eb442f7e81fbc5615bcf2b46c832

Since this is relatively similar to the work you've done with OpenThread.

carlescufi commented:
@Mani-Sadhasivam thanks for the patience. Can you update the `west.yml` commit so that it points to the tip of the current tree? We've fixed it now with @alexanderwachter.
https://github.com/zephyrproject-rtos/loramac-node/commits/master

carlescufi commented on west.yml:
```diff
@@ -88,6 +88,9 @@ manifest:
       revision: 9b591b289e1f37339bd038b5a1f0e6c8ad39c63a
       path: modules/lib/open-amp
+    - name: loramac-node
+      revision: 92f75bcafceab96473a38659d92080935ade354b
```
This should now be 29e516ec585b1a909af2b5f1c60d83e7d4d563e3

Mani-Sadhasivam commented:
@carlescufi Updated the PR!

rlubos commented on include/drivers/lora.h:
```diff
@@ -0,0 +1,136 @@
+		    data arrives.
+ * @return Length of the data received on success, negative on error
+ */
+static inline int lora_recv(struct device *dev, u8_t *data, s32_t timeout)
```
This API is missing a parameter that would indicate the size of the `data` buffer. How would the implementation know how many bytes can it copy into the `data` buffer? If we blindly copy data w/o taking buffer size into consideration we will sooner or later (but rather sooner) overwrite some data.

Mani-Sadhasivam replied:
I agree. I considered this while developing the API but I don't remember the reason why I decided to leave the size parameter :P

Mani-Sadhasivam replied:
@rlubos Okay, I got the reason now. The buffer copy happens with the size of the recevied length reported by the lora stack. So even if we pass the size parameter, we can't just copy those bytes since it may or may not be received. But there are couple of usecases with size parameter:

* We can check if the buffer size exceeded the max recv length i.e., 255
* Return error if we don't received the `size` bytes

rlubos replied:
@Mani-Sadhasivam The main reason for providing an extra `size` parameter is to detect when we can't copy the data so that we don't exceed the buffer provided. I'm not a Lora expert, so I'm not sure what would be the best action to take in case this happens, but the solutions I see:
* Return an error and drop the packet (results in data loss),
* Copy only the bytes that fit, return a number of bytes copied and drop the rest (that's how DATAGRAM sockets work),
* Copy only the bytes that fit, return a number of bytes copied and keep the rest of the packet for further reads (that's how STREAM packet works)

I understand though, that any of these will require some rework and would postpone this PR. Alternatively, we could specify in the API what's the size of the buffer that needs to be provided (as according to @alexanderwachter I understand that Lora packet has a maximum size, if it's 255 bytes used in the sample I guess it could also be a reasonable approach).

rlubos commented on drivers/lora/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library_named(loramac-node)
```
I think this declaration belongs to `loramac-node` repo. @SebastianBoe?

alexanderwachter replied:
This file is built first. We have to declare it here.

rlubos reviewed (commented):
@carlescufi Not exactly similar to OT, as we don't reuse the existing loramac-node module CMake build system, but I guess it was not as straightforward to integrate as OT was (at least it looks complex).

Left a few comments to consider, not necesirally CMake related.

carlescufi commented:
@Mani-Sadhasivam @alexanderwachter can we make 100% sure that the new LoRa sample is being built by CI before we merge this PR.

Mani-Sadhasivam commented:
@carlescufi Sure, we can wait. I waited for almost 4 months ;) Btw, there are few comments from @rlubos and @alexanderwachter, do you want me to incorporate them now or I can do it after merging this initial PR?

carlescufi commented:
@m

> @carlescufi Sure, we can wait. I waited for almost 4 months ;) Btw, there are few comments from @rlubos and @alexanderwachter, do you want me to incorporate them now or I can do it after merging this initial PR?

Please incorporate them now before merging this PR.

alexanderwachter commented:
@Mani-Sadhasivam I think you should incorporate the comments from @rlubos.
@rlubos The buffer has a fixed size. But a parameter would be nice.

jukkar commented on drivers/lora/Kconfig:
```diff
@@ -0,0 +1,30 @@
+# Top-level configuration file for LORA drivers.
+
+menuconfig LORA
+	bool "LoRa support"
```
It might be a good idea to mark the LoRa support as `[EXPERIMENTAL]` here so that it is easier to tweak the API if needed.

rlubos commented:
>@rlubos The buffer has a fixed size. But a parameter would be nice.

I understand, but there's no information whatsoever what is the expected buffer size in the API header file. So at a minimum, we should provide this information. Still, I think that extra `size_t` argument passed for in the function would be a neater solution.

alexanderwachter commented:
@rlubos i totally agree with the buffer size argument.

Mani-Sadhasivam commented:
@rlubos @alexanderwachter @carlescufi Incorporated the review comments. Please take a look

WilliamGFish commented:
@Mani-Sadhasivam This looks great but could your SX1276 not cover the whole SX12xx range? Obviously there may be work to make work generic.

The LoRaMac-Node  project also provides support for SX1272/73, SX1276/77/78/79 and SX1261/2 radios.

Mani-Sadhasivam commented:
> @Mani-Sadhasivam This looks great but could your SX1276 not cover the whole SX12xx range? Obviously there may be work to make work generic.
> 
> The LoRaMac-Node project also provides support for SX1272/73, SX1276/77/78/79 and SX1261/2 radios.

@WilliamGFish Of course! But I don't have any modules other than SX1276, so can't test. If others in the community has the modules and interest, it can always be extended. What I have provided is the base platform.
